src/sdb.c

```diff
@@ -454,58 +451,69 @@ SDB_API SdbList *sdb_foreach_list (Sdb* s) {
 	return list;
 }
 
-SDB_API int sdb_foreach (Sdb* s, SdbForeachCallback cb, void *user) {
+SDB_API bool sdb_foreach (Sdb* s, SdbForeachCallback cb, void *user) {
 	SdbListIter *iter;
 	char *k, *v;
 	SdbKv *kv;
-	if (!s) return 0;
+	bool found;
+	ut32 i;
+	if (!s) return false;
 	sdb_dump_begin (s);
 	while (sdb_dump_dupnext (s, &k, &v, NULL)) {
-		ut32 hash = sdb_hash (k);
-		SdbHashEntry *hte = ht_search (s->ht, hash);
-		if (hte) {
+		SdbKv *kv = ht_find_kvp (s->ht, k, &found);
+		if (found) {
 			free (k);
 			free (v);
-			kv = (SdbKv*)hte->data;
 			if (!*kv->value) {
 				// deleted = 1;
 				continue;
 			}
-			if (!cb (user, kv->key, kv->value))
-				return 0;
+			if (!cb (user, kv->key, kv->value)) {
+				return false;
+			}
 		} else {
 			int ret = cb (user, k, v);
 			free (k);
 			free (v);
-			if (!ret) return 0;
+			if (!ret) {
+				return false;
+			}
 		}
 	}
-	ls_foreach (s->ht->list, iter, kv) {
-		if (!kv->value || !*kv->value)
-			continue;
-		if (!cb (user, kv->key, kv->value))
-			return 0;
+	for (i = 0; i < s->ht->size; ++i) {
+		ls_foreach (s->ht->table[i], iter, kv) {
+			if (!kv->value || !*kv->value) {
+				continue;
+			} else if (!cb (user, kv->key, kv->value)) {
+				return false;
+			}
+		}
 	}
-	return 1;
+	return true;
 }
 
 // TODO: reuse sdb_foreach DEPRECATE WTF NOT READING THE CDB?
 SDB_API void sdb_list (Sdb* s) {
 	SdbListIter *iter;
 	SdbKv *kv;
+	ut32 i;
 	if (!s || !s->ht)
 		return;
-	ls_foreach (s->ht->list, iter, kv) {
-		if (!kv->value || !*kv->value)
-			continue;
-		printf ("%s=%s\n", kv->key, kv->value);
+	for (i = 0; i < s->ht->size; ++i) {
+		ls_foreach (s->ht->table[i], iter, kv) {
+			if (!kv->value || !*kv->value)
```

Comment:
maybe just invert the conditional to avoid having a continue in there? (if kv->value && *kv->value)